vllm/v1/engine/async_llm.py

```diff
@@ -536,6 +545,115 @@ async def abort(self, request_id: str | Iterable[str]) -> None:
+        async with self._pause_resume_lock:
+            async with self._pause_cond:
+                if self._is_paused:
+                    return {
+                        "paused": True,
+                        "mode": mode,
+                        "message": "Already paused",
+                        "aborted_requests": 0,
+                        "cache_cleared": False,
+                    }
+
+                self._is_paused = True
+
+            start_time = time.perf_counter()
+            aborted_requests = 0
+            cache_cleared = False
+
+            try:
+                if mode == "force":
+                    # Get all tracked request IDs directly from output_processor
+                    request_ids = list(self.output_processor.request_states.keys())
+                    aborted_requests = len(request_ids)
+                    if request_ids:
+                        await self.abort(request_ids)
+
+                # Wait for all requests to drain before clearing cache
+                while (
+                    self.output_processor.has_unfinished_requests()
+                    or self.engine_core.dp_engines_running()
+                ):
+                    await asyncio.sleep(0.05)
+
+                # Clear cache if requested
+                if clear_cache:
+                    await self.reset_prefix_cache()
+                    await self.reset_mm_cache()
+                    cache_cleared = True
+
+                elapsed = time.perf_counter() - start_time
+                return {
+                    "paused": True,
+                    "mode": mode,
+                    "message": "Generation paused successfully",
+                    "elapsed_seconds": elapsed,
+                    "aborted_requests": aborted_requests,
+                    "cache_cleared": cache_cleared,
+                }
+
+            except Exception:
+                # If pause fails, reset state to allow future requests
+                async with self._pause_cond:
+                    self._is_paused = False
+                    self._pause_cond.notify_all()
+                # Re-raise the exception so API returns HTTP 500 with error details
+                raise
+
+    async def resume_generation(self) -> dict[str, Any]:
+        """Resume generation after :meth:`pause_generation`."""
+
+        # Use lock to ensure atomicity with pause_generation
+        async with self._pause_resume_lock:
+            async with self._pause_cond:
+                if not self._is_paused:
+                    return {
+                        "paused": False,
+                        "message": "Not paused",
+                    }
+
+                self._is_paused = False
+                self._pause_cond.notify_all()  # Wake up all waiting requests
+
+            return {
+                "paused": False,
+                "message": "Generation resumed",
+            }
+
+    async def get_pause_status(self) -> dict[str, Any]:
```

Comment:
We should keep this simple, just define `is_paused` API. If you really need `num_unfinished_requests`, a separate API should be defined.

Reply:
agree.  The necessity of `num_unfinished_requests` is also discussed in https://github.com/vllm-project/vllm/issues/27901, which can be implemented in another PR.